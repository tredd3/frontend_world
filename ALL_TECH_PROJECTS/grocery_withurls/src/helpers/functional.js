export const pipe = (...args) => x => args.reduce((acc, fn) => fn(acc), x);

export const range = length => [...Array(length).keys()];

export const pad = (length, padCharacter) => value => {
  if (String(value).length >= length) return String(value);
  return range(length - String(value).length).map(() => padCharacter).join('') + String(value);
};

const make2digit = pad(2, '0');

export const formatDate = dateLike => {
  const date = new Date(dateLike);
  return `${date.getFullYear()}-${make2digit(date.getMonth() + 1)}-${make2digit(date.getDate())}`;
};

export const objectToQueryString = obj => Object.entries(obj)
  .reduce((acc, [key, value]) => [...acc, `${encodeURIComponent(key)}=${encodeURIComponent(Array.isArray(value)
    ? value.join(',') : value)}`], []).join('&');

export const queryStringToObject = qs => {
  if (!qs) return {};
  return qs.split('&')
    .map(aQsPart => aQsPart.split('=').map(decodeURIComponent))
    .reduce((acc, [name, value]) => {
      const convertedValue = value.includes(',') ? value.split(',') : [value];
      return { ...acc, [name]: convertedValue };
    }, {});
};
