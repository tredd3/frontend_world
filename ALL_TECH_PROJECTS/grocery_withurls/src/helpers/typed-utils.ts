export const flatMap = <T, U>(arr: T[], fn: ((x: T) => U[])): U[] => (
  arr.reduce(
    (acc: U[], x: T) => [...acc, ...fn(x)],
    [] as U[]
  )
);
