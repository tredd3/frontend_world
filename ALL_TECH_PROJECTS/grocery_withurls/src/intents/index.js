module.exports = process.env.NODE_ENV === 'development'
  ? require('./mock-intents')
  : require('./intents');
