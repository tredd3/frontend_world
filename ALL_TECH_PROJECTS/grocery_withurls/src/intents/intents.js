import promiseMemoize from 'promise-memoize';

const incomingMessageTypes = {
  receivedAccessToken: 'accessToken',
  receivedLocation: 'adparams'
};

const pendingPromises = {};

const isIOS = /iPad|iPhone|iPod/.test(navigator.userAgent) && !window.MSStream;

const handleIncomingMessage = type => data => {
  if (!pendingPromises[type]) return; // No pending promises waiting for this message
  const parsed = typeof data === 'string' ? JSON.parse(data) : data;

  if (type === incomingMessageTypes.receivedAccessToken) {
    pendingPromises[type].resolve({
      accessToken: parsed.access_tokens,
      refreshToken: parsed.refresh_token,
      apiKey: parsed.dataSetOne,
      expiresAt: Date.now() + (parsed.expires_in * 1000)
    });
  } else if (type === incomingMessageTypes.receivedLocation) {
    pendingPromises[type].resolve({
      lat: parsed.latitude,
      lng: parsed.longitude
    });
  }

  delete pendingPromises[type];
};

// The native app calls these global functions when it has to send us a message
window.sendAccessToken = handleIncomingMessage(incomingMessageTypes.receivedAccessToken);
window.sendAdParams = handleIncomingMessage(incomingMessageTypes.receivedLocation);

const sendToApp = payload => {
  const encodedData = window.btoa(JSON.stringify(payload));

  // eslint-disable-next-line no-underscore-dangle
  if (window.android && window.android.__externalCall) window.android.__externalCall(encodedData);
  // eslint-disable-next-line no-underscore-dangle
  if (window.__externalCall) window.__externalCall(encodedData);
  if (isIOS) window.webkit.messageHandlers.callback.postMessage(encodedData);
};

const sendToAppExpectingResponse = payload => {
  // payload.type must match incomingMessageTypes
  if (!pendingPromises[payload.type]) {
    let resolve;
    console.time(`Intent-${payload.type}`);
    const promise = new Promise(r => { resolve = r; });
    promise.then(data => {
      console.timeEnd(`Intent-${payload.type}`);
      return data;
    });
    pendingPromises[payload.type] = { promise, resolve };
  }

  sendToApp(payload);
  return pendingPromises[payload.type].promise;
};

export const getTokens = promiseMemoize(() => sendToAppExpectingResponse({
  type: 'accessToken',
  data: { getAccessToken: true }
}));

export const getLatLng = () => sendToAppExpectingResponse({
  type: 'adparams',
  data: { location: true }
}).then(({ lat, lng }) => {
  if (Number(lat) === 0 && Number(lng) === 0) throw new Error("Couldn't get location");
  return { lat, lng };
});

export const closeWindow = () => sendToApp({
  type: 'close',
  data: {}
});

export const makeCall = phoneNumber => sendToApp({
  type: 'makeCall',
  value: phoneNumber
});

export const showHeader = () => sendToApp({
  type: 'showHeader',
  data: { color: '#054484' }
});

export const hideHeader = () => sendToApp({
  type: 'hideHeader',
  data: { color: '#054484' }
});

export const chatWithCustomerService = () => sendToApp({
  type: 'launchbrowser',
  value: `https://wa.me/${917000370003}/?text=${encodeURIComponent('Hi')}`
});

// True if we want to give control to native for hardware back, false if we want web to have control
const toggleBackButtonControl = enableNativeControl => {
  if (window.android) {
    sendToApp({
      type: 'handleWebviewBack',
      value: enableNativeControl
    });
  }
};

export const startHandlingBackButton = callback => {
  if (typeof callback !== 'function') {
    return () => { /* No op */ };
  }

  toggleBackButtonControl(false);

  const stopHandlingBackButton = () => {
    toggleBackButtonControl(true);

    delete window.handleBackKey;
  };

  window.handleBackKey = callback;

  return stopHandlingBackButton;
};

export const notifyBannerClick = (bannerId, type, subType) => {
  if (window.AppInterface && window.AppInterface.onBannerClick) {
    window.AppInterface.onBannerClick(`app://jiokart/product/banner?type=${type}&subType=${subType}&bannerId=${bannerId}`);
    return true;
  }

  return false;
};
