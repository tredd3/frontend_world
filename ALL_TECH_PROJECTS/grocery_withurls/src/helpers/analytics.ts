import { getMobileOperatingSystem } from './utilities';
import { getUser } from '../services/user';
import { getCart, PaymentMethodId, paymentLabelById } from '../services/cart';

import { CheckoutData } from '../services/checkout';
import { CartPivot } from '../types';
import { flatMap } from './typed-utils';

let purchaseJourney: string;

export const setPurchaseJourney = (val: string) => {
  purchaseJourney = val;
};

export const getPurchaseJourney = () => purchaseJourney;

declare global {
  // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
  interface Window {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    digitalData: Record<string, any>;
    _satellite: {
      track: (event: string) => void;
    };
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const track = async (event: string, data: Record<string, any>) => {
  if (typeof window === 'undefined') return;
  const { userId } = await getUser();
  const userOS = getMobileOperatingSystem();
  const platform = userOS === 2 ? 'Android App' : userOS === 3 ? 'iOS App' : 'Web';

  const digitalData = {
    link: {
      ...data.link
    },
    internal: {
      ...data.internal
    },
    page: {
      pageInfo: {
        appName: 'JioMart Web'
      },
      ...data.page
    },
    user: {
      platform,
      loginStatus: 'Logged in',
      martId: userId,
      jioCenter: '',
      ...data.user
    },
    product: [
      ...(data.product ? data.product : [])
    ],
    interPromotion: {
      ...data.interPromotion
    },
    ...(data.product ? data.filterSection : {}),
    ...(data.purchase ? data.purchase : {})
  };

  window.digitalData = {
    ...window.digitalData,
    ...digitalData
  };

  // eslint-disable-next-line no-underscore-dangle
  if (window._satellite) window._satellite.track(event);
};

type LinkPositions = 'Top' | 'Middle' | 'Bottom';

export const trackLink = (linkName: string, linkType: string, linkPosition: LinkPositions) => (
  track('linkTracking', { link: { linkName, linkType, linkPosition } })
);

export const trackPage = (pageName: string) => (
  track('pageLoad', { page: { pageName } })
);

type FilterSection = {
  filterSelection: boolean;
  sortSelection: boolean;
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const trackProduct = (eventName: string, productData: Record<string, any>, filterSection?: FilterSection) => {
  const defaultFilterSection = {
    filterSelection: false,
    sortSelection: false
  };
  return track(eventName, {
    product: [productData],
    filterSection: (filterSection || defaultFilterSection)
  });
};

export const trackAPIShipments = (eventName: string, shipments: CheckoutData['Shipments'], purchase?: any) => {
  if (shipments && shipments.length) {
    track(eventName, {
      product: flatMap(shipments, shipment => shipment.Product)
        .map(product => ({
          productID: product.ProductSkuId,
          quantity: product.SelectedQuantity,
          price: product.SelectedQuantity * product.SP,
          productName: product.ProductName
        })),
      ...(purchase && { purchase })
    });
  }
};

type PurchaseArg = {
  paymentMethod: PaymentMethodId;
  paymentStatus?: 'Paid' | 'Unpaid';
  cart: CartPivot;
  failureReason?: string;
}

export const trackShipments = async (eventName: string, purchase?: PurchaseArg) => {
  const { shipments } = purchase?.cart || await getCart();
  if (!shipments || !shipments.length) return Promise.resolve();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let additionalInfo: Record<string, any> | null = null;

  if (purchase) {
    const { paymentMethod, cart, ...rest } = purchase;
    additionalInfo = {
      paymentMethod: paymentLabelById(paymentMethod),
      couponDiscount: cart.couponDiscountAmount,
      deliveryCharges: cart.deliveryCharge,
      totalOrderValue: cart.totalPayableAmount,
      transactionId: cart.cartId,
      ...rest
    };
  }

  return track(eventName, {
    product: flatMap(shipments, shipment => shipment.products)
      .map(product => ({
        productId: product.skuId,
        quantity: product.selectedQuantity,
        price: product.selectedQuantity * product.sp,
        productName: product.name
      })),
    ...(additionalInfo && additionalInfo)
  });
};
