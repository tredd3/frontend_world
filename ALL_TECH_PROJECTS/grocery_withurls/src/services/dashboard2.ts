import promiseMemoize from 'promise-memoize';
import { APIDashboardProduct } from '../types/api';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { DashboardProductPivot } from '../types';
import { pivotDashboardProduct } from './helpers/pivots';
import { getUser } from './user';

type Banner = { id: number; imageUrl: string };
type BannerWithTitle = Banner & { title: string };
type WidgetOfType<T> = { type: T; subType: number };
type BannerWidget<T> = WidgetOfType<T> & { banners: Banner[] };

export type HorizontalScrollableIcons = WidgetOfType<1> & { banners: BannerWithTitle[] };
export type Pager = BannerWidget<3>;
export type ThinCarouselBanner = BannerWidget<4>;
export type BigSquareCarouselBanner = BannerWidget<5>;
export type SmallSquareCarouselBanner = BannerWidget<6> & { title: string }
export type HorizontalProducts = WidgetOfType<7> & {
  title: string;
  viewAllLabel: string;
  products: DashboardProductPivot[];
};
export type W4x = WidgetOfType<11> & { banners: BannerWithTitle[] };
export type W9x = WidgetOfType<12> & { banners: BannerWithTitle[] } & { title: string };
export type W4xWithBanner = BannerWidget<15> & { title: string };
export type W9xWithBanner = BannerWidget<16> & { title: string };

export type Widget =
  HorizontalScrollableIcons
  | Pager
  | ThinCarouselBanner
  | BigSquareCarouselBanner
  | SmallSquareCarouselBanner
  | HorizontalProducts
  | W4x
  | W9x
  | W4xWithBanner
  | W9xWithBanner

type DashboardRequestType = {
  pincode?: number;
  storeId?: number;
};

type APIBanner = {
  bannerId: number;
  imageUrl: string;
};

type APIMiscellaneous = {
  name: string;
  mscId: number;
  imageUrl: string;
}

type APIWidget = {
  type: Widget['type'];
  subType: number;
  title: string;
  deepLinkText: string;
  data: {
    banners: null | APIBanner[];
    products: null | APIDashboardProduct[];
    miscellaneous: null | APIMiscellaneous[];
  };
};

export type DashboardResponseType = {
  widgets: APIWidget[];
};

const widgetBase = (widget: APIWidget) => ({
  type: widget.type,
  subType: widget.subType
});

const apiBannerToBanner = (banner: APIBanner | APIMiscellaneous): Banner => ({
  id: (banner as APIBanner).bannerId || (banner as APIMiscellaneous).mscId,
  imageUrl: banner.imageUrl
});

const createBanners = ({ data: widgetData }: APIWidget): { banners: Banner[] } => ({
  banners: ((widgetData.banners || widgetData.miscellaneous || []) as (APIBanner | APIMiscellaneous)[])
    .map(apiBannerToBanner)
});

const pivotBanner = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBanners(widget)
});

const createBannersWithTitle = (widget: APIWidget): { banners: BannerWithTitle[] } => ({
  banners: ((widget.data.miscellaneous || [])as APIMiscellaneous[])
    .map(banner => ({
      ...apiBannerToBanner(banner),
      title: banner.name
    }))
});

const createProducts = (widget: APIWidget): { products: DashboardProductPivot[] } => ({
  products: widget.data.products!.map(pivotDashboardProduct)
});

const pivotBannersWithTitle = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createBannersWithTitle(widget)
});

const pivotProducts = (widget: APIWidget) => ({
  ...widgetBase(widget),
  ...createProducts(widget)
});

export const pivotDashboardResponse = ({ widgets }: DashboardResponseType) => (
  widgets
    .filter(w => w.data.banners || w.data.miscellaneous || w.data.products)
    .map(widget => {
      switch (widget.type) {
        case 1: return pivotBannersWithTitle(widget) as HorizontalScrollableIcons;
        case 3: return pivotBanner(widget) as Pager;
        case 4: return pivotBanner(widget) as ThinCarouselBanner;
        case 5: return pivotBanner(widget) as BigSquareCarouselBanner;
        case 6:
          return {
            ...widgetBase(widget),
            title: widget.title,
            ...createBanners(widget)
          } as SmallSquareCarouselBanner;
        case 7:
          return {
            title: widget.title,
            viewAllLabel: widget.deepLinkText,
            ...pivotProducts(widget)
          } as HorizontalProducts;
        case 11: return pivotBannersWithTitle(widget) as W4x;
        case 12:
          return {
            title: widget.title,
            ...pivotBannersWithTitle(widget)
          } as W9x;
        case 15: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W4xWithBanner;
        case 16: return {
          ...widgetBase(widget),
          title: widget.title,
          ...createBanners(widget)
        } as W9xWithBanner;
        default: return null;
      }
    })
    .filter(<T>(x: T | null): x is T => x !== null)
);

export const getWidgetDefinitions = promiseMemoize(async () => {
  const { preferences: { pincode, storeId = 0 } } = await getUser();

  const apiWidgets = await makeApiCall<DashboardRequestType, DashboardResponseType>(apiRoutes.getDashboard, {
    body: { pincode, storeId }
  });

  return pivotDashboardResponse(apiWidgets);
}, { resolve: 'json' });
