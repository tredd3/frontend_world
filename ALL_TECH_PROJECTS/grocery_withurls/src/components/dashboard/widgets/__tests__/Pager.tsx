import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter, Router } from 'react-router-dom';
import { mocked } from 'ts-jest/utils';
import { createMemoryHistory } from 'history';
import PagerWidget from '../Pager';
import { pivotDashboardResponse, Pager, DashboardResponseType } from '../../../../services/dashboard2';
import mockDB from '../../../../mocks/db.json';
import * as analytics from '../../../../helpers/analytics';
import * as intents from '../../../../intents';

jest.mock('../../../../helpers/analytics');
jest.mock('../../../../intents');
jest.mock('../../../uiControls/Carousel', () => {
  const MockedCarousel: React.FC = ({ children }) => (
    <div>
      <div>Carousel placeholder</div>
      {children}
    </div>
  );

  return {
    __esModule: true,
    default: MockedCarousel
  };
});

const { setPurchaseJourney, track } = mocked(analytics);
const { notifyBannerClick } = mocked(intents);

const mockWidgetData = pivotDashboardResponse(mockDB.dashboard.Data as DashboardResponseType)
  .find(({ type }) => type === 3) as Pager;

describe('Pager widget tests', () => {
  beforeEach(() => {
    track.mockClear();
    notifyBannerClick.mockClear();
    setPurchaseJourney.mockClear();
  });

  it('should render the widget correctly', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <PagerWidget
          widgetData={mockWidgetData}
        />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it.each([true, false])('should handle click action correctly if AppInterface isDefined = %s', (isAppInterfaceDefined: boolean) => {
    const history = createMemoryHistory();
    notifyBannerClick.mockImplementation(() => isAppInterfaceDefined);
    const { getByTestId } = render(
      <Router history={history}>
        <PagerWidget
          widgetData={mockWidgetData}
        />
      </Router>
    );

    mockWidgetData.banners.forEach((_banner, index) => {
      fireEvent.click(getByTestId(`banner-link-${index}`));
      expect(track).toHaveBeenCalledTimes(index + 1);
      expect(track.mock.calls[index]).toMatchSnapshot();
      expect(setPurchaseJourney).toHaveBeenCalledTimes(index + 1);
      expect(setPurchaseJourney.mock.calls[index]).toMatchSnapshot();
      expect(notifyBannerClick.mock.calls[index]).toMatchSnapshot();
      expect(history.location.pathname + history.location.search).toMatchSnapshot();
    });

    notifyBannerClick.mockRestore();
  });
});
