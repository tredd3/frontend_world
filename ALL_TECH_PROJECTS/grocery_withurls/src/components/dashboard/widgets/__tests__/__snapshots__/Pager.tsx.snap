// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 1`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479738285,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 2`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 3`] = `
Array [
  1577479738285,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 4`] = `"/collections/1577479738285?type=3&subtype=1"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 5`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479742491,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 6`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 7`] = `
Array [
  1577479742491,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 8`] = `"/collections/1577479742491?type=3&subtype=1"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 9`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749430,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 10`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 11`] = `
Array [
  1577479749430,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 12`] = `"/collections/1577479749430?type=3&subtype=1"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 13`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749431,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 14`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 15`] = `
Array [
  1577479749431,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 16`] = `"/collections/1577479749431?type=3&subtype=1"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 17`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749432,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 18`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 19`] = `
Array [
  1577479749432,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 20`] = `"/collections/1577479749432?type=3&subtype=1"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 21`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749433,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 22`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 23`] = `
Array [
  1577479749433,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 24`] = `"/collections/1577479749433?type=3&subtype=1"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 25`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749434,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 26`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 27`] = `
Array [
  1577479749434,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = false 28`] = `"/collections/1577479749434?type=3&subtype=1"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 1`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479738285,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 2`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 3`] = `
Array [
  1577479738285,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 4`] = `"/"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 5`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479742491,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 6`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 7`] = `
Array [
  1577479742491,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 8`] = `"/"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 9`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749430,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 10`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 11`] = `
Array [
  1577479749430,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 12`] = `"/"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 13`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749431,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 14`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 15`] = `
Array [
  1577479749431,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 16`] = `"/"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 17`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749432,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 18`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 19`] = `
Array [
  1577479749432,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 20`] = `"/"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 21`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749433,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 22`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 23`] = `
Array [
  1577479749433,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 24`] = `"/"`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 25`] = `
Array [
  "bannerClicks",
  Object {
    "interPromotion": Object {
      "bannerId": 1577479749434,
      "widgetSubType": 1,
      "widgetType": 3,
    },
  },
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 26`] = `
Array [
  "Hero Banner",
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 27`] = `
Array [
  1577479749434,
  3,
  1,
]
`;

exports[`Pager widget tests should handle click action correctly if AppInterface isDefined = true 28`] = `"/"`;

exports[`Pager widget tests should render the widget correctly 1`] = `
<DocumentFragment>
  .emotion-2 {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  display: block;
  margin-bottom: 16px;
}

.emotion-2:hover,
.emotion-2:focus,
.emotion-2:visited,
.emotion-2:active {
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
}

.emotion-1 {
  background-color: transparent;
  margin: 16px 0;
}

.emotion-0 {
  border-radius: 3px;
}

<div>
    <div>
      Carousel placeholder
    </div>
    <a
      class="emotion-2"
      data-testid="banner-link-0"
      href="/collections/1577479738285?type=3&subtype=1"
    >
      <div
        class="emotion-1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; height: 170px;"
        >
          <img
            class="emotion-0"
            height="170"
            src="https://jiomerchant.akamaized.net/static/delivery_associate/A5NVC3OSTnBKHVGOj9KTRf6bcEYc+pGlNUZRXJgIFgY.png?impolicy=resizeforheight&h=170"
          />
        </span>
      </div>
    </a>
    <a
      class="emotion-2"
      data-testid="banner-link-1"
      href="/collections/1577479742491?type=3&subtype=1"
    >
      <div
        class="emotion-1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; height: 170px;"
        >
          <img
            class="emotion-0"
            height="170"
            src="https://jiomerchant.akamaized.net/static/delivery_associate/qckhQ2r5jQVdce7soFIrKZBRwEvuB5ZGw4qfAc7s8s.png?impolicy=resizeforheight&h=170"
          />
        </span>
      </div>
    </a>
    <a
      class="emotion-2"
      data-testid="banner-link-2"
      href="/collections/1577479749430?type=3&subtype=1"
    >
      <div
        class="emotion-1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; height: 170px;"
        >
          <img
            class="emotion-0"
            height="170"
            src="https://jiomerchant.akamaized.net/static/delivery_associate/0ww2kzGj3XegYcZ0w4WSBawbjjiwMspMg8owOIGNQNg.png?impolicy=resizeforheight&h=170"
          />
        </span>
      </div>
    </a>
    <a
      class="emotion-2"
      data-testid="banner-link-3"
      href="/collections/1577479749431?type=3&subtype=1"
    >
      <div
        class="emotion-1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; height: 170px;"
        >
          <img
            class="emotion-0"
            height="170"
            src="https://jiomerchant.akamaized.net/static/delivery_associate/Q+enZvTtMcqR1+7Kmx7xcsp63PoAwHQgZHKgbXR3M.png?impolicy=resizeforheight&h=170"
          />
        </span>
      </div>
    </a>
    <a
      class="emotion-2"
      data-testid="banner-link-4"
      href="/collections/1577479749432?type=3&subtype=1"
    >
      <div
        class="emotion-1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; height: 170px;"
        >
          <img
            class="emotion-0"
            height="170"
            src="https://jiomerchant.akamaized.net/static/delivery_associate/5HWANOwUueIKL97UiEDvUcYkmbppMnMGXdR6p5u9L3Q.png?impolicy=resizeforheight&h=170"
          />
        </span>
      </div>
    </a>
    <a
      class="emotion-2"
      data-testid="banner-link-5"
      href="/collections/1577479749433?type=3&subtype=1"
    >
      <div
        class="emotion-1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; height: 170px;"
        >
          <img
            class="emotion-0"
            height="170"
            src="https://jiomerchant.akamaized.net/static/delivery_associate/9+S0tjsbRQoaovmXpefEwuvGW+P5iriTsIfEreJKK0Q.png?impolicy=resizeforheight&h=170"
          />
        </span>
      </div>
    </a>
    <a
      class="emotion-2"
      data-testid="banner-link-6"
      href="/collections/1577479749434?type=3&subtype=1"
    >
      <div
        class="emotion-1"
      >
        <span
          class=" lazy-load-image-background "
          style="background-image: url(Placeholder.jpg); background-size: 100% 100%; color: transparent; display: inline-block; height: 170px;"
        >
          <img
            class="emotion-0"
            height="170"
            src="https://jiomerchant.akamaized.net/static/delivery_associate/pUQ9We+sNaCZogu84dM8roQzbPeLEVk1Hjoca9obDvM.png?impolicy=resizeforheight&h=170"
          />
        </span>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
